lib/auth/bot.go

```diff
@@ -0,0 +1,466 @@
+*/
+
+package auth
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/teleport/api/client/proto"
+	"github.com/gravitational/teleport/api/types"
+	apievents "github.com/gravitational/teleport/api/types/events"
+	"github.com/gravitational/teleport/lib/defaults"
+	"github.com/gravitational/teleport/lib/events"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+// BotResourceName returns the default name for resources associated with the
+// given named bot.
+func BotResourceName(botName string) string {
+	return "bot-" + strings.ReplaceAll(botName, " ", "-")
+}
+
+// createBotRole creates a role from a bot template with the given parameters.
+func createBotRole(ctx context.Context, s *Server, botName string, resourceName string, roleRequests []string) (types.Role, error) {
+	role, err := types.NewRole(resourceName, types.RoleSpecV5{
+		Options: types.RoleOptions{
+			// TODO: inherit TTLs from cert length?
+			MaxSessionTTL: types.Duration(12 * time.Hour),
+		},
+		Allow: types.RoleConditions{
+			Rules: []types.Rule{
+				// Bots read certificate authorities to watch for CA rotations
+				types.NewRule(types.KindCertAuthority, []string{types.VerbReadNoSecrets}),
+			},
+			Impersonate: &types.ImpersonateConditions{
+				Roles: roleRequests,
+			},
+		},
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	meta := role.GetMetadata()
+	meta.Description = fmt.Sprintf("Automatically generated role for bot %s", botName)
+	if meta.Labels == nil {
+		meta.Labels = map[string]string{}
+	}
+	meta.Labels[types.BotLabel] = botName
+	role.SetMetadata(meta)
+
+	err = s.UpsertRole(ctx, role)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return role, nil
+}
+
+// createBotUser creates a new backing User for bot use. A role with a
+// matching name must already exist (see createBotRole).
+func createBotUser(ctx context.Context, s *Server, botName string, resourceName string) (types.User, error) {
+	user, err := types.NewUser(resourceName)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	user.SetRoles([]string{resourceName})
+
+	metadata := user.GetMetadata()
+	metadata.Labels = map[string]string{
+		types.BotLabel:           botName,
+		types.BotGenerationLabel: "0",
+	}
+	user.SetMetadata(metadata)
+
+	// Traits need to be set to silence "failed to find roles or traits" warning
+	user.SetTraits(map[string][]string{
+		teleport.TraitLogins:     {},
+		teleport.TraitKubeUsers:  {},
+		teleport.TraitKubeGroups: {},
+	})
+
+	if err := s.CreateUser(ctx, user); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return user, nil
+}
+
+// createBot creates a new certificate renewal bot from a bot request.
+func (s *Server) createBot(ctx context.Context, req *proto.CreateBotRequest) (*proto.CreateBotResponse, error) {
```

Comment:
Should this function also validate that all roles specified in the CreateBotRequest are valid and exist in the system? Ignore if it's already validated somewhere and I just missed it.

Reply:
It doesn't, I'll add a quick check